Work in progress on composite support

roji opened the pull request:
This is work in progress on composite support for EFCore.PG 2.2, allowing you to map a CLR type to a PostgreSQL column defined as a composite type. This leverages composite support at the Npgsql level which is already there.

It's definitely not complete, but I wanted to get some feedback before going any deeper. It's probably a bit ambitious to push this for 2.2, but we may have enough value to deliver something.

What works:

* `ForNpgsqlHasComposite()` can be used to generate composite types in the database (much like `ForNpgsqlHasEnum()`. Currently there's no overload the infers the definition from the CLR type (like for enums).
* NpgsqlMigrationsSqlGenerator picks up the annotations and generates the types in the database (`CREATE TYPE ... AS (...)`).
* NpgsqlTypeMappingSource pulls global composite mappings from the ADO layer (like enums) and sets up new NpgsqlCompositeTypeMapping, including the configured NameTranslator. This requires exposing a new interface at the ADO level, so this depends on the unreleased 4.0.4.
* NpgsqlSqlTranslatingExpressionVisitor identifies MemberExpressions which traverse a composite type by finding a composite mapping for the CLR type. It then creates a new NpgsqlCompositeMemberExpression, including the composite's translated member name.
* NpgsqlCompositeMemberExpression is picked up by NpgsqlQuerySqlGenerator, which generates the appropriate member access SQL.
* All this allows you to query on composite fields (e.g. `WHERE my_entity.my_composite.some_field > 100`).
* Nested composites are (somewhat) supported. You can create them - the composite annotation contains an ordering vaue, allowing the nested composite to be created before the outer composite. Querying doesn't yet work though (see below).

What doesn't work:

* NpgsqlCompositeTypeMapping does *not* contain information about the composite's types. Unlike for enums, this cannot be inferred only from the CLR type, since the order of the composite's field is important and it only exists in the database. Unfortunately at the ADO level the actual database definition is only loaded when we actually connect to the database, but the EF Core layer looks at global composite mappings, which are set up without any specific databases. We'll have to think about what to do with this.
* This means that SQL and possibly code literals are unsupported at this time, because composite field knowledge is required at the mapping level. I *think* there's enough value without these for us to release this, but let me know what you think.
* Queries on fields of nested composites aren't translated yet, because `MemberAccessBindingExpressionVisitor.GetPropertyPath` does not traverse the outer composite. I think it's OK to move forward while documenting this limitation.
* Composite types currently cannot be altered in any way after creation (although this is supported by PostgreSQL). Again, not a blocker IMO.
* Scaffolding isn't yet implemented, although it shouldn't be difficult. This would mean scaffolding `ForNpgsqlHasComposite()`, but it will be up to the user to actually create the relevant CLR type and entity properties. This partial support is currently what we do for enums, maybe in the future we can actually generate the CLR types as well.
* No tests yet (tested ad-hoc)

@ajcvickers @divega @smitpatel you may be interested in this work which seems to push the limits a little.

austindrenski commented on src/EFCore.PG/Design/Internal/NpgsqlAnnotationCodeGenerator.cs:
```diff
@@ -88,6 +88,17 @@ public override MethodCallCodeFragment GenerateFluentApi(IModel model, IAnnotati
                         enumTypeDef.Schema, enumTypeDef.Name, enumTypeDef.Labels);
             }
 
+            if (annotation.Name.StartsWith(NpgsqlAnnotationNames.EnumPrefix))
```
I think this should be `CompositePrefix`.

austindrenski commented on src/EFCore.PG/Extensions/NpgsqlModelBuilderExtensions.cs:
```diff
@@ -247,6 +247,93 @@ public static class NpgsqlModelBuilderExtensions
+        /// <remarks>
+        /// See: https://www.postgresql.org/docs/current/rowtypes.html
+        /// </remarks>
+        /// <exception cref="ArgumentNullException">builder</exception>
```
`builder` isn't a parameter. Could you change this to a `<paramref name="modelBuilder" />`?

austindrenski commented on src/EFCore.PG/Extensions/NpgsqlModelBuilderExtensions.cs:
```diff
@@ -247,6 +247,93 @@ public static class NpgsqlModelBuilderExtensions
+            Check.NotNull(modelBuilder, nameof(modelBuilder));
+            Check.NotEmpty(name, nameof(name));
+            Check.NotNull(fields, nameof(fields));
+
```
How about adding `Check.NullButNotEmpty(schema, nameof(schema))`?

austindrenski commented on src/EFCore.PG/Extensions/NpgsqlModelBuilderExtensions.cs:
```diff
@@ -247,6 +247,93 @@ public static class NpgsqlModelBuilderExtensions
+        /// <remarks>
+        /// See: https://www.postgresql.org/docs/current/rowtypes.html
+        /// </remarks>
+        /// <exception cref="ArgumentNullException">builder</exception>
```
As above.

austindrenski commented on src/EFCore.PG/Metadata/PostgresComposite.cs:
```diff
@@ -0,0 +1,122 @@
+            [NotNull] string name,
+            [NotNull] params (string Name, string StoreType)[] fields)
+        {
+            if (FindPostgresComposite(annotatable, schema, name) is PostgresComposite CompositeType)
```
Should we add some `Check` preconditions? Or is this always called in a known context?

roji replied:
Not sure, but added anyway.

austindrenski commented on src/EFCore.PG/Metadata/PostgresComposite.cs:
```diff
@@ -0,0 +1,122 @@
+            [NotNull] string name)
+        {
+            Check.NotNull(annotatable, nameof(annotatable));
+            Check.NotEmpty(name, nameof(name));
```
As before, could we add `Check.NullButNotEmpty(schema, nameof(schema))`?

roji commented:
@austindrenski have pushed a commit addressing your comments thanks. Although this seems like it's too early for nits and other minor comments - I'm looking more for general comments on the approach, possible suggestions on how to handle the remaining issues, and opinions on whether this is viable with the limitations listed above (and for 2.2).

roji commented:
FYI there's a good chance we'll end up punting this for 3.0... @divega suggested an idea for working around the current lack of information in NpgsqlCompositeMapping, but there's a lot going on at the moment and 3.0 may change the query pipeline internals quite a bit in ways that may impact this. I'll still go over your review though, @austindrenski.

YohDeadfall commented:
It's a really large change, so I can't review it right now because of the presentation I mentioned. But definitely will do it in the nearest future.

Could you share @divega's suggestion with us?

roji commented:
> Could you share @divega's suggestion with us?

Sure, sorry about that, was in a bit of a hurry...

Basically the EF Core way of doing things is to manage as much as possible without a connection to the database. This is true when generating migrations, type mappings etc.

Now, when first doing the support for enums, my idea was to pull all enum mappings from the ADO layer, since in any case the user has to set up enum mappings there (`NpgsqlConnector.GlobalTypeMapper.MapEnum()`). Crucially, though, the global ADO mapping is enough to set up everything at the EF Core level as well: we get the CLR type, the database type and the name translator, and don't need anything else.

Composite types are a bit different - the global mapping set up at the ADO level isn't enough for us; although we can get the fields from the CLR type (just like the enum) and even use the translator to generate their database counterparts, we're missing their *order*, since the CLR property order will not necessarily correspond to the PostgreSQL composite type's field order. And the literal representation of a composite in PostgreSQL is positional - you need to specify values in the order of the composite fields (without names) - so you must know the order. At the ADO layer, the order is only known when the composite type is actually loaded from the database, when the first connection is instantiated and the types are loaded from pg_type and related tables - but the EF Core type mapper has no access to any specific connection.

So one way to solve this would be to somehow lazily communicate the database composite definition (with the order) to the EF Core type mapper/mapping on the first physical connection. That's complicated, and in theory you could have two databases with two different definitions being accessed with the same type mapper.

@divega's suggestion was simply to make sure the user communicates all the necessary information at the EF Core level (i.e. context options) rather than attempting to pull in information from the ADO layer. This is more in line with who EF Core works and should work fine, but has its own difficulties - the user would need to duplicate the information already mapped at the ADO layer (store type, CLR type, name translator). But it's definitely doable.

Then, for generating code literals (for data seeding) there are other challenges... To generate an instantiation of an arbitrary user CLR type we'd probably need to look for an appropriate constructor on the type (matching parameter names and types) and use that - not trivial but doable. As an alternative we could initialize the individual properties after the constructor (`new UserType() { X=1, Y=2}`), but I'm pretty sure the EF Core C# generator (CSharpHelper.cs) doesn't support the kind of expression yet (we could add it for 3.0 though).
